src/csharp/Grpc.Core/Internal/NativeCallbackDispatcher.cs

```diff
@@ -0,0 +1,101 @@
+#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Collections.Generic;
+using Grpc.Core.Logging;
+
+namespace Grpc.Core.Internal
+{
+    internal delegate int UniversalNativeCallback(IntPtr arg0, IntPtr arg1, IntPtr arg2, IntPtr arg3, IntPtr arg4, IntPtr arg5);
+
+    internal delegate int NativeCallbackDispatcherCallback(IntPtr tag, IntPtr arg0, IntPtr arg1, IntPtr arg2, IntPtr arg3, IntPtr arg4, IntPtr arg5);
+
+    internal class NativeCallbackDispatcher
+    {
+        static readonly ILogger Logger = GrpcEnvironment.Logger.ForType<NativeCallbackDispatcher>();
+        static readonly object staticLock = new object();
+
+        static NativeCallbackDispatcherCallback dispatcherCallback;
+
+        public static void Init(NativeMethods native)
+        {
+            lock (staticLock)
+            {
+                if (dispatcherCallback == null)
+                {
+                    dispatcherCallback = new NativeCallbackDispatcherCallback(HandleDispatcherCallback);
+                    native.grpcsharp_native_callback_dispatcher_init(dispatcherCallback);
+                }
+            }
+        }
+
+        public static NativeCallbackRegistration RegisterCallback(UniversalNativeCallback callback)
+        {
+            var gcHandle = GCHandle.Alloc(callback);
+            return new NativeCallbackRegistration(gcHandle);
+        }
+
+        private static UniversalNativeCallback GetCallback(IntPtr tag)
```

Comment:
nit: since it's only used in one place in here, suggest removing `GetCallback` and inlining the contents into `HandleDispatcherCallback` for readability

Reply:
done.